### DISCUSSION TITLE
How to use multiple exclusions in FileMatchingPatterns?

### DISCUSSION THREAD

**Discussion by roman-temekunidi** [microsoft/PR-Metrics#372] [Q&A] [answered]
I'm trying to exclude the `package-lock.json` and `packages.lock.json` files in `FileMatchingPatterns` . However, I can't make the exclusion work.

The following excludes `package-lock.json` correctly:
```yaml
FileMatchingPatterns: |
  **/*
  !**/package-lock.json
```

When I add the exclusion for `packages.lock.json`,  all package-lock.json files are no longer excluded:
```yaml
FileMatchingPatterns: |
  **/*
  !**/package-lock.json
  !**/packages.lock.json
```

I guess I'm missing something. Please advise.

**Comment by muiriswoulfe** [comment, marked as answer]
Hi @TemekRoman, thanks for using PR Metrics. I've taken a quick, provisional look and I think there might be an issue in the logic. I can definitely look into fixing this over the longer term. But I have a quick workaround for you now, to get you unblocked:

Using the following YAML should work:

```YAML
FileMatchingPatterns: |
  **/*
  !**/{package-lock.json,packages.lock.json}
```

Let me know how you get on with this.

**Reply by roman-temekunidi** [reply]
@muiriswoulfe 
Thanks for the swift reply. 
I've just checked the workaround you suggested, and it works as expected - both kinds of files are excluded.

If it's a bug, it is definitely worth fixing. Or at least update the docs to explain how to use multiple exclusions.

Thanks for your help!

**Reply by muiriswoulfe** [reply]
I agree that it's definitely not great. I've got a fix for this out now at #373, but it'll be next week before I'm able to push it out to the Marketplace.

**Reply by roman-temekunidi** [reply]
In the [overview.md](https://github.com/microsoft/PR-Metrics/blob/c207f96749814054c0ab3cc0e7eb7a1d00d23bf3/src/overview.md?plain=1#LL101C1-L101C88), there is a link to AzureDevOps [File matching patterns reference](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/file-matching-patterns?view=azure-devops) that explains how to use inclusions/exclusions. There is a section that features the [double exclude](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/file-matching-patterns?view=azure-devops#doubleexcl_examples).

Do you plan to support the double exclude as well?

**Reply by muiriswoulfe** [reply]
Makes sense. I've made some more updates to support this now.

**Reply by roman-temekunidi** [reply]
Awesome! Thanks a lot.

**Reply by muiriswoulfe** [reply]
The change is now available in v1.5.2, which was released a short while ago. Azure DevOps will likely pick up the latest version for you automatically but, if not, you can download it from the Marketplace.
